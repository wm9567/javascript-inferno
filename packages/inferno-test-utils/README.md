# inferno-test-utils

> Suite of utilities for testing Inferno applications

## Install

```
npm install inferno --save
npm install inferno-test-utils --save-dev
```

## Contents

* [`renderIntoDocument`](#renderintodocumentvnodetree)
* [`findAllInRenderedTree`](#findallinrenderedtreerenderedtree-predicate)
* [`findAllInVNodeTree`](#findallinvnodetreevnodetree-predicate)
* [`scryRenderedDOMElementsWithClass`](#scryrendereddomelementswithclassrenderedtree-classname)
* [`findRenderedDOMElementWithClass`](#findrendereddomelementwithclassrenderedtree-classname)
* [`scryRenderedDOMElementsWithTag`](#scryrendereddomelementswithtagrenderedtree-tagname)
* [`findRenderedDOMElementWithTag`](#findrendereddomelementwithtagrenderedtree-tagname)
* [`scryRenderedVNodesWithType`](#scryrenderedvnodeswithtyperenderedtree-type)
* [`findRenderedVNodeWithType`](#findrenderedvnodewithtyperenderedtree-type)
* [`scryVNodesWithType`](#scryvnodeswithtypevnodetree-type)
* [`findVNodeWithType`](#findvnodewithtypevnodetree-type)
* [`isVNode`](#isvnodeinstance)
* [`isVNodeOfType`](#isvnodeoftypeinstance-type)
* [`isDOMVNode`](#isdomvnodeinstance)
* [`isDOMVNodeOfType`](#isdomvnodeoftypeinstance-type)
* [`isFunctionalVNode`](#isfunctionalvnodeinstance)
* [`isFunctionalVNodeOfType`](#isfunctionalvnodeoftypeinstance-type)
* [`isClassVNode`](#isclassvnodeinstance)
* [`isClassVNodeOfType`](#isclassvnodeoftypeinstance-type)
* [`isDOMElement`](#isdomelementinstance)
* [`isDOMElementOfType`](#isdomelementoftypeinstance-type)
* [`isRenderedClassComponent`](#isrenderedclasscomponentinstance)
* [`isRenderedClassComponentOfType`](#isrenderedclasscomponentoftypeinstance-type)

## Usage

### `renderIntoDocument(VNodeTree)`

Renders `VNodeTree` into a detached DOM element in the document and returns a rendered `VNode` tree.

_This function requires a DOM_.

```js
const VNodeTree = (
  <div className="outer">
    <SomeComponent className="inner"/>
  </div>
);
const renderedTree = renderIntoDocument(VNodeTree);
```

### `findAllInRenderedTree(renderedTree, predicate)`

Calls `predicate` with each `VNode` instance in `renderedTree`. 

Returns an array of `VNodes` where `predicate` returns `true`.

```js
const VNodeTree = (
  <div className="outer">
    <SomeComponent className="inner"/>
  </div>
);
const renderedTree = renderIntoDocument(VNodeTree);
const predicate = (VNode) => VNode.type === SomeComponent;
const result = findAllInRenderedTree(renderedTree, predicate);
```

### `findAllInVNodeTree(VNodeTree, predicate)`

Calls `predicate` with each `VNode` instance in `VNodeTree`. 

Returns an array of `VNodes` where `predicate` returns `true`.

```js
const VNodeTree = (
  <div className="outer">
    <SomeComponent className="inner"/>
  </div>
);
const predicate = (VNode) => VNode.type === SomeComponent;
const result = findAllInVNodeTree(VNodeTree, predicate);
```

### `scryRenderedDOMElementsWithClass(renderedTree, className)`

Returns an array of DOM elements with `className`.

`className` can be a space-separated string or an array of strings.

```js
const VNodeTree = (
  <div className="outer">
    <SomeComponent className="inner one"/>
    <SomeComponent className="inner two"/>
  </div>
);
const renderedTree = renderIntoDocument(VNodeTree);
const result1 = scryRenderedDOMElementsWithClass(renderedTree, 'inner');
const result2 = scryRenderedDOMElementsWithClass(renderedTree, 'inner one');
const result3 = scryRenderedDOMElementsWithClass(renderedTree, ['inner', 'two']);
const result4 = scryRenderedDOMElementsWithClass(renderedTree, 'three'); // Empty array
```

### `findRenderedDOMElementWithClass(renderedTree, className)`

Returns a single DOM element with `className`. If more than one matches are found, throws an error.

`className` can be a space-separated string or an array of strings.

```js
const VNodeTree = (
  <div className="outer">
    <SomeComponent className="inner one"/>
    <SomeComponent className="inner two"/>
  </div>
);
const renderedTree = renderIntoDocument(VNodeTree);
const result1 = scryRenderedDOMElementsWithClass(renderedTree, 'outer');
const result2 = scryRenderedDOMElementsWithClass(renderedTree, 'inner one');
// Will throw an error because more than 1 matches were found...
const result3 = scryRenderedDOMElementsWithClass(renderedTree, 'inner');
```

### `scryRenderedDOMElementsWithTag(renderedTree, tagName)`

Returns an array of DOM elements with `tagName`.

```js
const VNodeTree = (
  <div>
    <h1>Heading</h1>
    <p>Paragraph One</p>
    <p>Paragraph Two</p>
    <p>Paragraph Three</p>
  </div>
);
const renderedTree = renderIntoDocument(VNodeTree);
const result1 = scryRenderedDOMElementsWithTag(renderedTree, 'h1');
const result3 = scryRenderedDOMElementsWithTag(renderedTree, 'p');
const result4 = scryRenderedVNodesWithType(renderedTree, 'span'); // Empty array
```

### `findRenderedDOMElementWithTag(renderedTree, tagName)`

Returns a single DOM element with `tagName`. If more than one matches are found, throws an error.

```js
const VNodeTree = (
  <div>
    <h1>Heading</h1>
    <div>
      <p>Paragraph One</p>
      <p>Paragraph Two</p>
      <p>Paragraph Three</p>
    </div>
  </div>
);
const renderedTree = renderIntoDocument(VNodeTree);
const result1 = findRenderedDOMElementWithTag(renderedTree, 'h1');
// Will throw an error because more than 1 matches were found...
const result2 = findRenderedDOMElementWithTag(renderedTree, 'p');
```

### `scryRenderedVNodesWithType(renderedTree, type)`

Returns an array of rendered `VNodes` with `type`.

```js
const VNodeTree = (
  <div>
    <h1>Heading</h1>
    <SomeComponent/>
    <SomeComponent/>
  </div>
);
const renderedTree = renderIntoDocument(VNodeTree);
const result1 = scryRenderedVNodesWithType(renderedTree, 'h1');
const result2 = scryRenderedVNodesWithType(renderedTree, SomeComponent);
const result3 = scryRenderedVNodesWithType(renderedTree, 'p'); // Empty array
```

### `findRenderedVNodeWithType(renderedTree, type)`

Returns a single rendered `VNode` with `type`. If more than one matches are found, throws an error.

```js
const VNodeTree = (
  <div>
    <h1>Heading</h1>
    <p>Paragraph 1</p>
    <p>Paragraph 2</p>
    <SomeComponent/>
    <AnotherComponent/>
    <AnotherComponent/>
  </div>
);
const renderedTree = renderIntoDocument(VNodeTree);
const result1 = findRenderedVNodeWithType(renderedTree, 'h1');
const result2 = findRenderedVNodeWithType(renderedTree, SomeComponent);
// Will throw an error because more than 1 matches were found...
const result3 = findRenderedVNodeWithType(renderedTree, 'p');
const result4 = findRenderedVNodeWithType(renderedTree, AnotherComponent);
```

### `scryVNodesWithType(VNodeTree, type)`

Returns an array of `VNodes` with `type`.

```js
const VNodeTree = (
  <div>
    <h1>Heading</h1>
    <SomeComponent/>
    <SomeComponent/>
  </div>
);
const result1 = scryVNodesWithType(VNodeTree, 'h1');
const result2 = scryVNodesWithType(VNodeTree, SomeComponent);
const result3 = scryVNodesWithType(VNodeTree, 'p'); // Empty array
```

### `findVNodeWithType(VNodeTree, type)`

Returns a single `VNode` with `type`. If more than one matches are found, throws an error.

```js
const VNodeTree = (
  <div>
    <h1>Heading</h1>
    <SomeComponent/>
    <SomeComponent/>
  </div>
);
const result1 = findVNodeWithType(VNodeTree, 'h1');
// Will throw an error because more than 1 matches were found...
const result2 = findVNodeWithType(VNodeTree, SomeComponent);
```

### `isVNode(instance)`

Returns `true` when `instance` is a `VNode`.

### `isVNodeOfType(instance, type)`

Returns `true` when `instance` is a `VNode` of `type`.

### `isDOMVNode(instance)`

Returns `true` when `instance` is a DOM-type `VNode`.

### `isDOMVNodeOfType(instance, type)`

Returns `true` when `instance` is a DOM-type `VNode` of `type`.

### `isFunctionalVNode(instance)`

Returns `true` when `instance` is a functional-type `VNode`.

### `isFunctionalVNodeOfType(instance, type)`

Returns `true` when `instance` is a functional-type `VNode` of `type`.

### `isClassVNode(instance)`

Returns `true` when `instance` is a class-type `VNode`.

### `isClassVNodeOfType(instance, type)`

Returns `true` when `instance` is a class-type `VNode` of `type`.

### `isDOMElement(instance)`

Returns `true` when `instance` is a DOM element.

### `isDOMElementOfType(instance, type)`

Returns `true` when `instance` is a DOM element of `type`.

### `isRenderedClassComponent(instance)`

Returns `true` when `instance` is a rendered class `VNode`.

### `isRenderedClassComponentOfType(instance, type)`

Returns `true` when `instance` is a rendered class `VNode` of `type`.
