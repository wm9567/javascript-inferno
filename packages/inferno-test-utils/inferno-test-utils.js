module.exports = require('inferno/dist/inferno-test-utils.node');
module.exports.default = module.exports;