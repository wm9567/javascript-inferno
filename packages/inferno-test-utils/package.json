{
  "name": "inferno-test-utils",
  "version": "0.8-alpha",
  "license": "MIT",
  "description": "Provides test utilities to help with creating and maintaining tests for Inferno applications",
  "author": {
    "name": "Dominic Gannaway",
    "email": "dg@domgan.com"
  },
  "bugs": {
    "url": "https://github.com/trueadm/inferno/issues"
  },
  "homepage": "https://github.com/trueadm/inferno#readme",
  "keywords": [
    "babel",
    "react",
    "inferno",
    "framework",
    "interfaces",
    "user interfaces"
  ],
  "dependencies": {
    "inferno": "0.8-alpha"
  },
  "main": "inferno-test-utils.js",
  "repository": "https://github.com/trueadm/inferno"
}
