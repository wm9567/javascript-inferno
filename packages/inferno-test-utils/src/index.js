import {
	findAllInRenderedTree,
	findRenderedDOMComponentWithClass,
	isCompositeComponent,
	isCompositeComponentWithType,
	isDOMComponent,
	isDOMComponentElement,
	isElement,
	isElementOfType,
	renderIntoDocument,
	scryRenderedDOMComponentsWithClass,
	scryRenderedDOMComponentsWithTag,
	scryRenderedComponentsWithType,
	findRenderedDOMComponentWithTag,
	findRenderedComponentWithType,
	mockComponent
} from '../../../build/tools/testUtils';

export {
	renderIntoDocument,
	isElement,
	isElementOfType,
	isDOMComponent,
	isDOMComponentElement,
	isCompositeComponent,
	isCompositeComponentWithType,
	findAllInRenderedTree,
	scryRenderedDOMComponentsWithClass,
	scryRenderedDOMComponentsWithTag,
	scryRenderedComponentsWithType,
	findRenderedDOMComponentWithClass,
	findRenderedDOMComponentWithTag,
	findRenderedComponentWithType,
	mockComponent
};

export default {
	renderIntoDocument,
	isElement,
	isElementOfType,
	isDOMComponent,
	isDOMComponentElement,
	isCompositeComponent,
	isCompositeComponentWithType,
	findAllInRenderedTree,
	scryRenderedDOMComponentsWithClass,
	scryRenderedDOMComponentsWithTag,
	scryRenderedComponentsWithType,
	findRenderedDOMComponentWithClass,
	findRenderedDOMComponentWithTag,
	findRenderedComponentWithType,
	mockComponent,
};
