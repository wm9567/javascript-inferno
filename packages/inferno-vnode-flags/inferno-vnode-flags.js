module.exports = require('./build');
module.exports.default = module.exports;
