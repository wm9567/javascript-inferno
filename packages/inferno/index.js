/**
 * Created by ADM on 31/12/15.
 */
