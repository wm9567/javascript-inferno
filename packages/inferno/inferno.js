module.exports = require('./dist/inferno');
