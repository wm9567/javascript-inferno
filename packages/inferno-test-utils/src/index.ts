import { InfernoChildren, InfernoInput, render, VNode } from 'inferno';
import Component from 'inferno-component';
import createElement from 'inferno-create-element';
import { isArray, isFunction, isNumber, isObject, isString, throwError } from 'inferno-shared';
import VNodeFlags from 'inferno-vnode-flags';

// Type Checkers

export function isVNode(instance: any): boolean {
	return Boolean(instance) && isObject(instance) &&
		isNumber(instance.flags) && instance.flags > 0;
}

export function isVNodeOfType(instance: VNode, type: string | Function): boolean {
	return isVNode(instance) && instance.type === type;
}

export function isDOMVNode(instance: VNode): boolean {
	return isVNode(instance) && isString(instance.type);
}

export function isDOMVNodeOfType(instance: VNode, type: string): boolean {
	return isDOMVNode(instance) && instance.type === type;
}

export function isFunctionalVNode(instance: VNode): boolean {
	return isVNode(instance) && Boolean(instance.flags & VNodeFlags.ComponentFunction);
}

export function isFunctionalVNodeOfType(instance: VNode, type: Function): boolean {
	return isFunctionalVNode(instance) && instance.type === type;
}

export function isClassVNode(instance: VNode): boolean {
	return isVNode(instance) && Boolean(instance.flags & VNodeFlags.ComponentClass);
}

export function isClassVNodeOfType(instance: VNode, type: Function): boolean {
	return isClassVNode(instance) && instance.type === type;
}

export function isDOMElement(instance: any): boolean {
	return Boolean(instance) && isObject(instance) &&
		instance.nodeType === 1 && isString(instance.tagName);
}

export function isDOMElementOfType(instance: any, type: string): boolean {
	return isDOMElement(instance) && isString(type) &&
		instance.tagName.toLowerCase() === type.toLowerCase();
}

export function isRenderedClassComponent(instance: any): boolean {
	return Boolean(instance) && isObject(instance) && isVNode(instance._vNode) &&
		isFunction(instance.render) && isFunction(instance.setState);
}

export function isRenderedClassComponentOfType(instance: any, type: Function): boolean {
	return isRenderedClassComponent(instance) &&
		isFunction(type) && instance._vNode.type === type;
}

// Render Utilities

class Wrapper extends Component<any, any> {
	render() {
		return this.props.children;
	}
}

export function renderIntoDocument(input: InfernoInput): InfernoChildren {
	const wrappedInput = createElement(Wrapper, null, input);
	const parent = document.createElement('div');
	document.body.appendChild(parent);
	return render(wrappedInput, parent);
}

// Recursive Finder Functions

export function findAllInRenderedTree(renderedTree: any, predicate: (vNode: VNode) => boolean): VNode[] {
	if (isRenderedClassComponent(renderedTree)) {
		return findAllInVNodeTree(renderedTree._lastInput, predicate);
	} else {
		throwError('findAllInRenderedTree(renderedTree, predicate) renderedTree must be a rendered class component');
	}
}

export function findAllInVNodeTree(vNodeTree: VNode, predicate: (vNode: VNode) => boolean): VNode[] {
	if (isVNode(vNodeTree)) {
		let result: VNode[] = predicate(vNodeTree) ? [ vNodeTree ] : [];
		const children: any = vNodeTree.children;

		if (isRenderedClassComponent(children)) {
			result = result.concat(findAllInVNodeTree(children._lastInput, predicate));

		} else if (isVNode(children)) {
			result = result.concat(findAllInVNodeTree(children, predicate));

		} else if (isArray(children)) {
			children.forEach((child) => {
				result = result.concat(findAllInVNodeTree(child, predicate));
			});
		}
		return result;
	} else {
		throwError('findAllInVNodeTree(vNodeTree, predicate) vNodeTree must be a VNode instance');
	}
}

// Finder Helpers

function parseSelector(filter) {
	if (isArray(filter)) {
		return filter;
	} else if (isString(filter)) {
		return filter.trim().split(/\s+/);
	} else {
		return [];
	}
}

function findOneOf(tree: any, filter: any, name: string, finder: Function): any {
	const all = finder(tree, filter);
	if (all.length > 1) {
		throwError(`Did not find exactly one match (found ${all.length}) for ${name}: ${filter}`);
	} else {
		return all[ 0 ];
	}
}

// Scry Utilities

export function scryRenderedDOMElementsWithClass(renderedTree: any, classNames: string | string[]): Element[] {
	return findAllInRenderedTree(renderedTree, (instance) => {
		if (isDOMVNode(instance)) {
			let domClassName = instance.dom.className;
			if (!isString(domClassName)) { // SVG, probably
				domClassName = instance.dom.getAttribute('class') || '';
			}
			const domClassList = parseSelector(domClassName);
			return parseSelector(classNames).every((className) => {
				return domClassList.indexOf(className) !== -1;
			});
		}
		return false;
	}).map((instance) => instance.dom);
}

export function scryRenderedDOMElementsWithTag(renderedTree: any, tagName: string): Element[] {
	return findAllInRenderedTree(renderedTree, (instance) => {
		return isDOMVNodeOfType(instance, tagName);
	}).map((instance) => instance.dom);
}

export function scryRenderedVNodesWithType(renderedTree: any, type: string | Function): VNode[] {
	return findAllInRenderedTree(renderedTree, (instance) => isVNodeOfType(instance, type));
}

export function scryVNodesWithType(vNodeTree: VNode, type: string | Function): VNode[] {
	return findAllInVNodeTree(vNodeTree, (instance) => isVNodeOfType(instance, type));
}

// Find Utilities

export function findRenderedDOMElementWithClass(renderedTree: any, classNames: string | string[]): Element {
	return findOneOf(renderedTree, classNames, 'class', scryRenderedDOMElementsWithClass);
}

export function findRenderedDOMElementWithTag(renderedTree: any, tagName: string): Element {
	return findOneOf(renderedTree, tagName, 'tag', scryRenderedDOMElementsWithTag);
}

export function findRenderedVNodeWithType(renderedTree: any, type: string | Function): VNode {
	return findOneOf(renderedTree, type, 'component', scryRenderedVNodesWithType);
}

export function findVNodeWithType(vNodeTree: VNode, type: string | Function): VNode {
	return findOneOf(vNodeTree, type, 'VNode', scryVNodesWithType);
}

export default {

	isVNode,
	isVNodeOfType,

	isDOMVNode,
	isDOMVNodeOfType,

	isFunctionalVNode,
	isFunctionalVNodeOfType,

	isClassVNode,
	isClassVNodeOfType,

	isDOMElement,
	isDOMElementOfType,

	isRenderedClassComponent,
	isRenderedClassComponentOfType,

	renderIntoDocument,

	findAllInRenderedTree,
	findAllInVNodeTree,

	scryRenderedDOMElementsWithClass,
	findRenderedDOMElementWithClass,

	scryRenderedDOMElementsWithTag,
	findRenderedDOMElementWithTag,

	scryRenderedVNodesWithType,
	findRenderedVNodeWithType,

	scryVNodesWithType,
	findVNodeWithType
};
