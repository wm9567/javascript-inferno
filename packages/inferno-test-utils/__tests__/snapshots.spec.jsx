import { Component, Fragment, createFragment } from 'inferno';
import { renderToSnapshot } from 'inferno-test-utils';
import { ChildFlags } from 'inferno-vnode-flags';

describe('Snapshots', () => {
  class Foobar extends Component {
    render({ children }) {
      return <div className="Testing">{children}</div>;
    }
  }

  function Testing({ children }) {
    return (
      <span>
        {children}
        Extra
      </span>
    );
  }

  describe('JSX', () => {
    if (window.usingJest) {
      it('Should render attributes and className', () => {
        expect(renderToSnapshot(<Foobar />)).toMatchSnapshot();
      });

      it('Should render component children', () => {
        expect(
          renderToSnapshot(
            <Foobar>
              <div className="ok">ABC</div>
            </Foobar>
          )
        ).toMatchSnapshot();
      });

      it('Should render html element', () => {
        expect(
          renderToSnapshot(
            <a onClick={() => {}} aria-colspan="3" className="foo">
              Bar
            </a>
          )
        ).toMatchSnapshot();
      });

      it('Should render multiple elements', () => {
        expect(
          renderToSnapshot(
            <Testing>
              <a>T</a>
              <span className="foo">est</span>
              12
            </Testing>
          )
        ).toMatchSnapshot();
      });

      it('Should render deeper component', () => {
        class ComP extends Component {
          render() {
            return (
              <div>
                Okay
                {this.props.children}
                <span className="yea" />
              </div>
            );
          }
        }

        expect(
          renderToSnapshot(
            <div className="okay" aria-rowindex={1} onClick={() => console.log}>
              <ComP>
                <div>
                  <ul>
                    <li>Okay</li>
                  </ul>
                </div>
              </ComP>
            </div>
          )
        ).toMatchSnapshot();
      });

      it('Should render fragment', () => {
        const fragmentA = createFragment([<div id="a1">A1</div>, <div>A2</div>], ChildFlags.HasNonKeyedChildren, 'A');

        const fragmentB = createFragment([<div id="b1">B1</div>], ChildFlags.HasNonKeyedChildren, 'B');

        const fragmentC = createFragment([<div id="c1">C1</div>, <div>C2</div>, <div>C3</div>], ChildFlags.HasNonKeyedChildren, 'C');

        expect(
          renderToSnapshot(
            <div>
              {fragmentA}
              {fragmentB}
              {fragmentC}
            </div>
          )
        ).toMatchSnapshot();
      });

      it('Should render fragment root', () => {
        expect(
          renderToSnapshot(
            <Fragment>
              <div>First</div>
              <div>Second</div>
            </Fragment>
          )
        ).toMatchSnapshot();
      });

      it('Should render fragment from component root', () => {
        class Comp extends Component {
          render() {
            return <>{this.props.children}</>;
          }
        }

        expect(
          renderToSnapshot(
            <Comp>
              <div>1</div>
            </Comp>
          )
        ).toMatchSnapshot();
      });

      it('Should not fail when returning children array from component root, which contains text node. Github #1404', () => {
        const Label = ({ label, htmlFor, children, optional = false, ...props }) => {
          if (optional && !label) {
            return children;
          }
          return (
            <>
              <label {...props} htmlFor={htmlFor}>
                {label}
              </label>
              {children}
            </>
          );
        };
        expect(renderToSnapshot(<Label>{[<span>o</span>, <span>k</span>, 'asd', 1, null, false, true, void 0]}</Label>)).toMatchSnapshot();
      });

      it('Should not fail when returning children array from component root, Github #1404', () => {
        const Label = ({ label, htmlFor, children, optional = false, ...props }) => {
          if (optional && !label) {
            return children;
          }

          return (
            <>
              <label {...props} htmlFor={htmlFor}>
                {label}
              </label>
              {children}
            </>
          );
        };

        expect(renderToSnapshot(<Label>{[<span>o</span>, <span>k</span>]}</Label>)).toMatchSnapshot();
      });

      it('Should flush setStates before building snapshot', () => {
        class App extends Component {
          constructor(props) {
            super(props);

            this.state = {
              foo: ''
            };
          }

          componentDidMount() {
            this.setState({
              foo: '##BAR##'
            });
          }

          render() {
            return (
              <div className="App">
                <header className="App-header">
                  <img src="logo" className="App-logo" alt="logo" />
                  <p>
                    Edit <code>src/App.js</code> and save to reload.
                  </p>
                  <a className="App-link" href="https://reactjs.org" target="_blank" rel="noopener noreferrer">
                    {this.state.foo}
                  </a>
                </header>
              </div>
            );
          }
        }

        expect(renderToSnapshot(<App />)).toMatchSnapshot();
      });
    }
  });
});
