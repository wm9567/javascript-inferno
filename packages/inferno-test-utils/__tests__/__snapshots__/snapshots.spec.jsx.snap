// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots JSX Should flush setStates before building snapshot 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo"
    />
    <p>
      Edit 
      <code>
        src/App.js
      </code>
       and save to reload.
    </p>
    <a
      className="App-link"
      href="https://reactjs.org"
      rel="noopener noreferrer"
      target="_blank"
    >
      ##BAR##
    </a>
  </header>
</div>
`;

exports[`Snapshots JSX Should not fail when returning children array from component root, Github #1404 1`] = `
Array [
  <label />,
  <span>
    o
  </span>,
  <span>
    k
  </span>,
]
`;

exports[`Snapshots JSX Should not fail when returning children array from component root, which contains text node. Github #1404 1`] = `
Array [
  <label />,
  <span>
    o
  </span>,
  <span>
    k
  </span>,
  "asd",
  "1",
]
`;

exports[`Snapshots JSX Should render attributes and className 1`] = `
<div
  className="Testing"
/>
`;

exports[`Snapshots JSX Should render component children 1`] = `
<div
  className="Testing"
>
  <div
    className="ok"
  >
    ABC
  </div>
</div>
`;

exports[`Snapshots JSX Should render deeper component 1`] = `
<div
  aria-rowindex={1}
  className="okay"
  onClick={[Function]}
>
  <div>
    Okay
    <div>
      <ul>
        <li>
          Okay
        </li>
      </ul>
    </div>
    <span
      className="yea"
    />
  </div>
</div>
`;

exports[`Snapshots JSX Should render fragment 1`] = `
<div>
  Array [
    <div
      id="a1"
    >
      A1
    </div>,
    <div>
      A2
    </div>,
  ]
  Array [
    <div
      id="b1"
    >
      B1
    </div>,
  ]
  Array [
    <div
      id="c1"
    >
      C1
    </div>,
    <div>
      C2
    </div>,
    <div>
      C3
    </div>,
  ]
</div>
`;

exports[`Snapshots JSX Should render fragment from component root 1`] = `
<div>
  1
</div>
`;

exports[`Snapshots JSX Should render fragment root 1`] = `
Array [
  <div>
    First
  </div>,
  <div>
    Second
  </div>,
]
`;

exports[`Snapshots JSX Should render html element 1`] = `
<a
  aria-colspan="3"
  className="foo"
  onClick={[Function]}
>
  Bar
</a>
`;

exports[`Snapshots JSX Should render multiple elements 1`] = `
<span>
  <a>
    T
  </a>
  <span
    className="foo"
  >
    est
  </span>
  12
  Extra
</span>
`;
